fix(coffee): don't allow duplicate repositories

tareknaser360 opened the pull request:
- added a hashset of the remote repositories names to the storage information of coffee on disk (JSON)
- added a function to check if the remote repository clone was deleted from disk manually and called it when needed

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -34,31 +36,39 @@ use crate::CoffeeArgs;
 pub struct CoffeStorageInfo {
     pub config: config::CoffeeConf,
     pub repositories: Vec<RepositoryInfo>,
+    pub repos_names: HashSet<String>,
```
```suggestion
    pub repositories: HasMap<String, RepositoryInfo>,
```

Why not this?

tareknaser360 replied:
Yes this actually makes more sense. Thanks

tareknaser360 commented:
I changed the code to use a `HashMap<String, RepositoryInfo>` to store repository information.
I also refactored the `rm_remote` function to use the hashmap.

vincenzopalazzo reviewed (commented):
Looks good to me from a fist scan, I should test it locally tomorrow, I put it on my todo list.

P.S: at the end of may I promise you to implement a test framework for it :)

tareknaser360 commented:
> P.S: at the end of may I promise you to implement a test framework for it :)

Great! I would love to help with that.

vincenzopalazzo reviewed (commented):
We are almost there, but this PR only addresses minor issues and does not solve the root problem. Even if we remove the `.coffee/storage.json` file (which is necessary when the JSON format changes), the refresh function still does not work, and the error persists.

My proposal is to:
- add a method to coffee called `nurse`, which will remove all the repositories in the `.coffee/repository` folder that are not in the storage. This will also clean up the plugins in the coffee configuration at the end, and it is not a trivial operation. 
- Additionally, we should run the `refresh_remote` function inside the `inventory` method, but without modifying the list of repositories. If there are any inconsistencies, we should return an error message stating that `the coffee storage is out of sync and that the user should run coffee nurse` to fix the problem.

With these changes, we should be able to solve our problem. Great job so far! Only now have I been able to get a clear idea of what needs to be done.

What do you think?

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -288,50 +295,66 @@ impl PluginManager for CoffeeManager {
+    async fn refresh_remotes(&mut self) -> Result<(), CoffeeError> {
+        self.repos.retain(|_, repo| {
+            let repo_path = repo.url().path_string;
+            Path::new(&repo_path).exists()
```
but I do not want that the remote sync will modify this without user permission (maybe it is running coffee in the wrong directory) so we want only return the error if there is some inconsistency and tell the user how to fix it

tareknaser360 replied:
I made it log a warning and return an error but not modify any files or folders.

tareknaser360 commented on coffee_core/src/coffee.rs:
```diff
@@ -288,50 +295,66 @@ impl PluginManager for CoffeeManager {
+    }
+
     async fn add_remote(&mut self, name: &str, url: &str) -> Result<(), CoffeeError> {
+        self.refresh_remotes().await?;
```
The problem with that is if we do, we can't call the `nurse` method because the error will be returned first and the clean up logic will not run.

tareknaser360 commented:
I added a new command `nurse`
it does 2 things:
- check for repositories in `self.repos` and if one doesn't exist on disk, deletes it from storage.
- check for files and folders in `.coffee.repositories` and deletes any unrelated file (that doesn't exist in `self.repos`)

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -288,50 +297,71 @@ impl PluginManager for CoffeeManager {
         self.storage.store(&self.storage_info()).await
     }
 
+    async fn remote_sync(&mut self) -> Result<(), CoffeeError> {
```
```suggestion
    async fn remote_sync(&mut self) -> Result<(), CoffeeError> {
```

can you call this command `nurse`?

tareknaser360 replied:
The command is called nurse. This is just a function to check for the local folder that we call with every other command.

vincenzopalazzo commented on coffee_cmd/src/cmd.rs:
```diff
@@ -53,6 +53,9 @@ pub enum CoffeeCommand {
     #[clap(arg_required_else_help = true)]
     Show { plugin: String },
+    /// clean up remote repositories storage information
+    #[clap(arg_required_else_help = false)]
```
```suggestion
    #[clap(arg_required_else_help = true)]
```

tareknaser360 replied:
Should coffee nurse command take any arguments?

vincenzopalazzo replied:
Ah now I got it, no your are right!

vincenzopalazzo commented on coffee_cmd/src/main.rs:
```diff
@@ -98,6 +98,17 @@ async fn main() -> Result<(), CoffeeError> {
             Err(err) => Err(err),
         },
+        CoffeeCommand::Nurse {} => {
+            let mut spinner = term::spinner(&format!("Refreshing storage"));
```
```suggestion
            let mut spinner = term::spinner(&format!("Trying restoring from a corrupted status!"));
```

We should also check if there is a corrupted status, if not we return a message that there is no need to run the command

tareknaser360 replied:
I modified the nurse function to return a json object with the status of the file and print a message to the user accordingly.

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -257,9 +264,10 @@ impl PluginManager for CoffeeManager {
     }
 
     async fn list(&mut self, remotes: bool) -> Result<CoffeeList, CoffeeError> {
+        self.remote_sync().await?;
```
why not put this method inside the `inventory` method of the plugin manager?

tareknaser360 replied:
But then it will be called with coffee nurse itself and the error will be returned before the code runs

vincenzopalazzo replied:
Right, there is no way to check what kind of command the inventory is running?

tareknaser360 replied:
I searched and there was a way using an extra Boolean but this is cleaner in my opinion.

vincenzopalazzo replied:
lets keep this and think about a possible solution in the mean while

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -288,50 +297,71 @@ impl PluginManager for CoffeeManager {
     }
 
     async fn rm_remote(&mut self, name: &str) -> Result<(), CoffeeError> {
+        self.remote_sync().await?;
```
yeah I think we need to find another place where to call this method

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -358,6 +389,34 @@ impl PluginManager for CoffeeManager {
+                if entry.file_type()?.is_dir() {
+                    fs::remove_dir_all(entry.path())?;
+                    log::debug!("folder removed: {}", entry_name);
+                } else if entry.file_type()?.is_file() {
+                    fs::remove_file(entry.path())?;
+                    log::debug!("file removed: {}", entry_name);
+                }
```
add a comment here that if there is a missing file inside the repository we considered the repository corrupted, it is not trivial for the people that will go to read the  code

vincenzopalazzo commented on coffee_core/src/lib.rs:
```diff
@@ -14,6 +14,7 @@ pub enum CoffeeOperation {
     /// Setup(core lightning root path)
     Setup(String),
     Show(String),
+    Nurse(),
```
```suggestion
    Nurse,
```

this is also valid rust right?

tareknaser360 replied:
Yes it is.

vincenzopalazzo reviewed (changes requested):
Some comments but the code looks really good!

I like that you put the PR when it is ready to be review, thanks

vincenzopalazzo commented on coffee_cmd/src/main.rs:
```diff
@@ -98,6 +98,23 @@ async fn main() -> Result<(), CoffeeError> {
+            let mut spinner = term::spinner(&format!("Trying restoring from a corrupted status!"));
+            match coffee.nurse().await {
+                Ok(val) => {
+                    let status = val["nurse"].as_str().unwrap();
+                    if status == "corrupt" {
+                        spinner.message("Storage refreshed");
```
Please make a model and not return a plan json here

tareknaser360 replied:
Sure. Just pushed the code with the changes.

vincenzopalazzo commented on coffee_cmd/src/main.rs:
```diff
@@ -98,6 +98,23 @@ async fn main() -> Result<(), CoffeeError> {
+            let mut spinner = term::spinner(&format!("Trying restoring from a corrupted status!"));
+            match coffee.nurse().await {
+                Ok(val) => {
+                    let status = val.status;
+                    if status == "corrupt" {
```
```suggestion
                    let status = val.status;
                    if status == "corrupt" {
```

It is fine, but the Rust way of doing this (also the functional way to do this, not just in Rust) is to define a type and match on this type.

So let's imagine having an enum like the following one in our model

```rust
pub enum NurseStatus {
      Currupted,
     Sane,
}
```

You than you do the match like 

```rust
match val.status {
    NurseStatus::Corrupted => {},
    NurseStatus::Sane => {},
}
```

This improves the code a lot

tareknaser360 replied:
Yes this is much cleaner

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -288,50 +297,71 @@ impl PluginManager for CoffeeManager {
+                        .iter()
+                        .position(|elem| elem.name() == *plugin.name())
+                    {
+                        let plugin_name = &plugins[ind].name()[..];
```
```suggestion
                        let plugin_name = &plugins[ind].name().clone()
```

Do not work in this way?

tareknaser360 replied:
Yes this works

tareknaser360 commented:
0b8daeb solves #123

vincenzopalazzo commented on coffee_core/src/coffee.rs:
```diff
@@ -395,13 +394,15 @@ impl PluginManager for CoffeeManager {
+
+        // delete any unrelated files or folders in `repositories` folder
+        let repos_path = PathBuf::from(format!("{}/repositories/", self.config.root_path));
         for entry in fs::read_dir(repos_path)? {
```
we should use tokio as `fs` otherwise you break the async workflow

vincenzopalazzo reviewed (changes requested):
ok we are super close, just a minor change in the API that you are using

tareknaser360 commented:
This is now rebased on top of master.
